qualtran/qref_interop/_bloq_to_qref.py

```diff
@@ -0,0 +1,311 @@
+            "Soquets referencing more than one index in composite register are not yet supported. "
+            f"The error was caused by the following soquet: {soquet}."
+        )
+    # We add another suffix iff soquet references idx in composite register
+    suffix = f"_{soquet.idx[0]}" if soquet.idx else ""
+    return (
+        # If soquet references BlogqInstance, the corresponding object in Bartiq
+        # references child - construct dotted relative name.
+        # Otherwise, soquet references the parent port, and so for the output direction,
+        # the port is in_parent_port, which is why we include _opposte here.
+        f"{_bloq_instance_name(soquet.binst)}.{direction}_{soquet.reg.name}{suffix}"
+        if isinstance(soquet.binst, BloqInstance)
+        else f"{_opposite(direction)}_{soquet.reg.name}{suffix}"
+    )
+
+
+def _import_connection(connection: QualtranConnection) -> dict[str, Any]:
+    """Import connection from Qualtran."""
+    return {
+        "source": _relative_port_name(connection.left, "out"),
+        "target": _relative_port_name(connection.right, "in"),
+    }
+
+
+def _import_resources(bloq: Bloq) -> list[dict[str, Any]]:
+    """Import resources from Bloq's t_complexity method."""
+    t_complexity = bloq.t_complexity()
+    resource_names = ["t", "clifford", "rotations"]
+    resources = []
+    for name in resource_names:
+        resources.append(
+            {
+                "name": name,
+                "value": _ensure_primitive_type(getattr(t_complexity, name)),
+                "type": "additive",
+            }
+        )
+    return resources
+
+
+def _ensure_primitive_type(value: Any) -> Union[int, float, str, None]:
+    """Ensure given value is of primitive type (e.g. is not a sympy expression)."""
+    return value if value is None or isinstance(value, (int, float, str)) else str(value)
+
+
+def _extract_input_params(bloq: Bloq, ports: list[PortV1], local_variables) -> list[str]:
+    """Extracts input_params from bloq's t_complexity and port sizes."""
+    params_from_t_complexity = _extract_input_params_from_t_complexity(bloq)
+    params_from_ports = _extract_input_params_from_ports(ports)
+    params = list(set(params_from_t_complexity + params_from_ports) - set(local_variables))
+    return sorted(params)
+
+
+def _extract_input_params_from_t_complexity(bloq: Bloq) -> list[str]:
+    input_params = set()
+    t_complexity = bloq.t_complexity()
+    resource_names = ["t", "clifford", "rotations"]
+    for name in resource_names:
+        try:
+            input_params.update(getattr(t_complexity, name).free_symbols)
+        except AttributeError:
+            pass
+    return [str(symbol) for symbol in input_params]
+
+
+def _extract_input_params_from_ports(ports) -> list[str]:
+    input_params = set()
+    for port in ports:
+        input_params = input_params | sympy.sympify(port.size).free_symbols
+
+    return [str(param) for param in input_params]
```

Comment:
Can you please add docstrings to these two methods?

Reply:
Done!